### PR TITLE
JCL-419: harden e2e code

### PR DESCRIPTION
**Pull request by timea-solid**
We saw lately too many CI problems. On investigation, it turned out that the problem occurs because the tests do not do a proper cleanup.
A solution here is to also clean up before starting the tests (making sure the containers are empty for example).

It still happens that some tests randomly fail, I am not sure if we can simply up the retry number...

### CODE REVIEW

File: integration/base/src/main/java/com/inrupt/client/integration/base/AccessGrantScenarios.java [hunk 1]
```diff
@@ -154,17 +152,19 @@ static void setup() throws IOException {
             podUrl += Utils.FOLDER_SEPARATOR;
         }
 
-
         privateContainerURI = URIBuilder.newBuilder(URI.create(podUrl))
                 .path(State.PRIVATE_RESOURCE_PATH + "/").build();
 
+        authResourceOwnerClient = createAuthenticatedClient();
+        //if a tests fails it can be that the cleanup was not properly done, so we do it here too
+        Utils.deleteContentsRecursively(authResourceOwnerClient, privateContainerURI);
```
Comment (acoburn) [inline integration/base/src/main/java/com/inrupt/client/integration/base/AccessGrantScenarios.java:160]: We can't do this deletion here, because it will cause any other concurrently-running test to fail.

File: integration/base/src/main/java/com/inrupt/client/integration/base/AuthenticationScenarios.java [hunk 2]
```diff
@@ -125,23 +121,31 @@ static void setup() {
...
+                .path("auth-test-" + UUID.randomUUID())
                 .build();
 
             publicContainerURI = URIBuilder.newBuilder(URI.create(podUrl))
                     .path(PUBLIC_RESOURCE_PATH + "/").build();
 
-            createContainer(publicContainerURI);
-            prepareACR(publicContainerURI);
+            //if a tests fails it can be that the cleanup was not properly done, so we do it here too
+            Utils.deleteContentsRecursively(localAuthClient, publicContainerURI);
```
Comment (acoburn) [inline integration/base/src/main/java/com/inrupt/client/integration/base/AuthenticationScenarios.java:147]: As mentioned elsewhere, we can't delete this data here, because it will interfere with other concurrently-running tests.

File: integration/base/src/main/java/com/inrupt/client/integration/base/AuthenticationScenarios.java [hunk 3]
```diff
@@ -159,31 +163,28 @@ static void setup() {
 
         privateContainerURI = URIBuilder.newBuilder(URI.create(podUrl))
                 .path(PRIVATE_RESOURCE_PATH + "/").build();
+        //if a tests fails it can be that the cleanup was not properly done, so we do it here too
```
Comment (acoburn) [inline integration/base/src/main/java/com/inrupt/client/integration/base/AuthenticationScenarios.java:166]: If _tests_ fail, the cleanup (`@AfterAll`) gets run, it's only if the setup (`@BeforeAll`) fails that the cleanup doesn't happen.

File: integration/base/src/main/java/com/inrupt/client/integration/base/AuthenticationScenarios.java [hunk 4]
```diff
@@ -159,31 +163,28 @@ static void setup() {
...
+        Utils.deleteContentsRecursively(localAuthClient, privateContainerURI);
```
Comment (acoburn) [inline integration/base/src/main/java/com/inrupt/client/integration/base/AuthenticationScenarios.java:167]: We don't want to delete data in any setup phase, as it may interfere with other tests.

File: integration/base/src/main/java/com/inrupt/client/integration/base/CoreModulesResource.java [hunk 5]
```diff
@@ -130,22 +130,29 @@ static void setup() {
...
                 .build();
 
             publicContainerURI = URIBuilder.newBuilder(URI.create(podUrl))
                     .path(PUBLIC_RESOURCE_PATH + "/").build();
-            createAuthenticatedClient();
-            createContainer(publicContainerURI);
-            prepareACR(publicContainerURI);
+
+            //if a tests fails it can be that the cleanup was not properly done, so we do it here too
+            Utils.deleteContentsRecursively(localAuthClient, publicContainerURI);
```
Comment (acoburn) [inline integration/base/src/main/java/com/inrupt/client/integration/base/CoreModulesResource.java:154]: ditto

File: integration/base/src/main/java/com/inrupt/client/integration/base/DomainModulesResource.java [hunk 6]
```diff
@@ -122,22 +121,30 @@ static void setup() {
...
+                .path("domain-test-" + UUID.randomUUID())
                 .path(testContainer)
                 .build();
 
             publicContainerURI = URIBuilder.newBuilder(URI.create(podUrl))
                     .path(PUBLIC_RESOURCE_PATH + FOLDER_SEPARATOR).build();
-            createContainer(publicContainerURI);
-            prepareACR(publicContainerURI);
+            //if a tests fails it can be that the cleanup was not properly done, so we do it here too
+            Utils.deleteContentsRecursively(localAuthClient, publicContainerURI);
```
Comment (acoburn) [inline integration/base/src/main/java/com/inrupt/client/integration/base/DomainModulesResource.java:146]: ditto

File: integration/base/src/main/java/com/inrupt/client/integration/base/Utils.java [hunk 7]
```diff
@@ -166,6 +173,96 @@ public static Set<Triple> publicAgentPolicyTriples(final URI acl) {
         return triples;
     }
 
+    public static void createPublicContainer(final SolidSyncClient authClient, final URI publicContainerURI) {
+        if (!exists(authClient, publicContainerURI)) {
+            try (SolidContainer newContainer = new SolidContainer(publicContainerURI)) {
+                try (final SolidContainer container = authClient.create(newContainer)) {
+                    container.getMetadata().getAcl().ifPresent(acl -> {
+                        int i = 0; // in case we cannot read the acl, we try 2 times before giving up
```
Comment (acoburn) [inline integration/base/src/main/java/com/inrupt/client/integration/base/Utils.java:181]: This issue this method is fixing is (I believe) symptomatic of some other issue, and it would be better to fix the underlying issue.

For example, the ACL is not present because another process has just deleted the corresponding resource.

File: integration/base/src/main/java/com/inrupt/client/integration/base/Utils.java [hunk 8]
```diff
@@ -166,6 +173,96 @@ public static Set<Triple> publicAgentPolicyTriples(final URI acl) {
...
+        }
+    }
+
+    static void deleteContentsRecursively(final SolidSyncClient client, final URI url) {
+        if (exists(client, url)) {
+            deleteRecursive(client, url, new AtomicInteger(0));
+        }
+    }
+
+    static void deleteRecursive(final SolidSyncClient client, final URI url, final AtomicInteger depth) {
```
Comment (acoburn) [inline integration/base/src/main/java/com/inrupt/client/integration/base/Utils.java:226]: I would avoid the use of a shared AtomicInteger here; instead, you can just have a `final int depth` and perform arithmetic operations on that in the recursive calls

Comment (edwardsph) [reply integration/base/src/main/java/com/inrupt/client/integration/base/Utils.java:226]: Yes, the AtomicInteger was useful in an asynchronous context

**Review by edwardsph — COMMENTED** [review]
Taking a step back and looking at how you manage test containers, this is what I see:

If `PUBLIC_RESOURCE_PATH` doesn't exist then you have
`  publicTestContainerURI = podUrl/test-UUID/`
and you rely on the ACR of the pod root being right.

If it does exist, then you have
`  publicTestContainerURI = podUrl/PUBLIC_RESOURCE_PATH/test-UUID/`
Then you create `podUrl/PUBLIC_RESOURCE_PATH/` and set up the ACR and either tear down this container at the end or try and ensure it is empty at the start. Since this is a fixed location, it will clash with other test runs.

Why not make that container unique too so you create `podUrl/PUBLIC_RESOURCE_PATH-UUID/`. That would isolate your tests but still use a defined prefix to make the test container identifiable (if that was the reason for allowing someone to set `PUBLIC_RESOURCE_PATH`. I may not have understood why you want someone to define the test container outside the tests.

The downside of my approach is that if something goes wrong with teardown, you can be left with old test containers but they will not interfere with anything and you can have a separate housekeeping process to clean them out. However, I think you will find that once the tests are more stable you will have less of a problem with this.

File: integration/base/src/main/java/com/inrupt/client/integration/base/CoreModulesResource.java [hunk 9]
```diff
@@ -117,50 +113,42 @@ static void setup() {
...
                 .build();
 
-            publicContainerURI = URIBuilder.newBuilder(URI.create(podUrl))
-                    .path(PUBLIC_RESOURCE_PATH + "/").build();
-            createAuthenticatedClient();
-            createContainer(publicContainerURI);
-            prepareACR(publicContainerURI);
-        }
+        //if a tests fails it can be that the cleanup was not properly done, so we do it here too
+        Utils.deleteContentsRecursively(localAuthClient, publicContainerURI);
```
Comment (edwardsph) [inline integration/base/src/main/java/com/inrupt/client/integration/base/CoreModulesResource.java:142]: This line becomes redundant since you are now defining a unique new  `publicContainerURI`, so it doesn't yet exist and cannot have any contents. It is created on the next line. There may be other cases like this to check for.

File: integration/base/src/main/java/com/inrupt/client/integration/base/Utils.java (hunk 8 above)
Comment (timea-solid) [reply integration/base/src/main/java/com/inrupt/client/integration/base/Utils.java:228]: However, I cannot use it as `final int depth` but I do need a final in the lambda later. I did not find a better solution.

File: integration/base/src/main/java/com/inrupt/client/integration/base/AccessGrantScenarios.java [hunk 10]
```diff
@@ -479,6 +449,10 @@ void accessGrantGetRdfTest(final Session resourceOwnerSession, final Session req
         final SolidSyncClient resourceOwnerClient = SolidSyncClient.getClientBuilder()
             .build().session(resourceOwnerSession);
 
+        final URI testRDFresourceURI = URIBuilder.newBuilder(privateContainerURI)
+                .path("resource-accessGrantGetRdfTest.ttl")
+                .build();
+
         try (final SolidRDFSource resource = new SolidRDFSource(testRDFresourceURI, null, null)) {
             assertDoesNotThrow(() -> resourceOwnerClient.create(resource));
```
Comment (NSeydoux) [inline integration/base/src/main/java/com/inrupt/client/integration/base/AccessGrantScenarios.java:457]: A pattern I typically use with the JS client for such case, where the resource URI is irrelevant to the client (but still needs to be tracked) is to get the server to create the resource by POSTing to the parent container, which URL the client does care about (here that would be `privateContainerURI`). Is this a pattern that would be possible with the JCL, or does the target URL needs to be known before creating a resource?

Comment (acoburn) [reply integration/base/src/main/java/com/inrupt/client/integration/base/AccessGrantScenarios.java:457]: In the high-level client, we exclusively use PUT for create and update. POST and PATCH are usable directly in the low-level client.

Comment (NSeydoux) [reply integration/base/src/main/java/com/inrupt/client/integration/base/AccessGrantScenarios.java:457]: It's absolutely not critical for GA, but would something like `createChild` make sense in the `SolidClient` ?

**Review by acoburn — APPROVED** [review]
